spynnaker/pyNN/models/neuron/synaptic_manager.py

```diff
@@ -592,120 +606,575 @@ def _write_synaptic_matrix_and_master_population_table(
+        # Write the data to generate a matrix on-chip
         max_row_info = self.__get_max_row_info(
             synapse_info, post_vertex_slice, app_edge, machine_time_step)
+        is_undelayed = bool(max_row_info.undelayed_max_n_synapses)
+        is_delayed = bool(max_row_info.delayed_max_n_synapses)
+
+        # Create initial master population table entries if a single
+        # matrix is to be created for the whole application vertex
+        syn_block_addr = 0xFFFFFFFF
+        if is_undelayed and app_key_info is not None:
+            block_addr, syn_block_addr = self.__reserve_mpop_block(
+                block_addr, max_row_info.undelayed_max_bytes,
+                max_row_info.undelayed_max_words, app_key_info,
+                all_syn_block_sz, app_edge.pre_vertex.n_atoms,
+                post_vertex_slice, synapse_info)
+            syn_max_addr = block_addr
+            key = (app_edge, synapse_info, post_vertex_slice.lo_atom)
+            self.__app_edge_info[key] = (
+                syn_block_addr, max_row_info.undelayed_max_words,
+                block_addr - syn_block_addr)
+        elif app_key_info is not None:
+            self.__poptable_type.add_invalid_entry(
+                app_key_info.key_and_mask, app_key_info.core_mask,
+                app_key_info.core_shift, app_key_info.n_neurons)
+        delay_block_addr = 0xFFFFFFFF
+        if is_delayed and delay_app_key_info is not None:
+            block_addr, delay_block_addr = self.__reserve_mpop_block(
+                block_addr, max_row_info.delayed_max_bytes,
+                max_row_info.delayed_max_words,
+                delay_app_key_info, all_syn_block_sz,
+                app_edge.pre_vertex.n_atoms * app_edge.n_delay_stages,
+                post_vertex_slice, synapse_info)
+            delay_max_addr = block_addr
+            key = (app_edge, synapse_info, post_vertex_slice.lo_atom)
+            self.__delay_edge_info[key] = (
+                delay_block_addr, max_row_info.delayed_max_words,
+                block_addr - delay_block_addr)
+        elif delay_app_key_info is not None:
+            self.__poptable_type.add_invalid_entry(
+                delay_app_key_info.key_and_mask, delay_app_key_info.core_mask,
+                delay_app_key_info.core_shift, delay_app_key_info.n_neurons)
+
+        for m_edge in m_edges:
+            syn_mat_offset = syn_block_addr
+            d_mat_offset = delay_block_addr
+            pre_idx = graph_mapper.get_machine_vertex_index(m_edge.pre_vertex)
+            pre_slice = graph_mapper.get_slice(m_edge.pre_vertex)
+
+            # Write the information needed to generate delays
+            self.__write_on_chip_delay_data(
+                max_row_info, app_edge, pre_slices, pre_idx, post_slices,
+                post_slice_index, pre_slice, post_vertex_slice, synapse_info,
+                machine_time_step)
+
+            if is_undelayed and app_key_info is not None:
```

Comment:
i honestly think these ifs for un delayed and delayed should be seperate functions. Just to make thismethod readable

Reply:
Even better - made a single function that both can use.